vignettes/decorate-modules-output.Rmd

```diff
@@ -0,0 +1,389 @@
+
+
+```{r}
+tm_decorated_plot <- function(label = "module", transforms = list(), decorators = teal_transform_module()) {
+  module(
+    label = label,
+    ui = function(id, decorators) {
+      ns <- NS(id)
+      div(
+        selectInput(ns("dataname"), label = "select dataname", choices = NULL),
+        selectInput(ns("x"), label = "select x", choices = NULL),
+        selectInput(ns("y"), label = "select y", choices = NULL),
+        ui_transform_data(ns("decorate"), transforms = decorators),
+        plotOutput(ns("plot")),
+        verbatimTextOutput(ns("text"))
+      )
+    },
+    server = function(id, data, decorators) {
+      moduleServer(id, function(input, output, session) {
+        observeEvent(data(), {
+          updateSelectInput(inputId = "dataname", choices = names(data()))
+        })
+
+        observeEvent(input$dataname, {
+          req(input$dataname)
+          updateSelectInput(inputId = "x", choices = colnames(data()[[input$dataname]]))
+          updateSelectInput(inputId = "y", choices = colnames(data()[[input$dataname]]))
+        })
+
+        plot_data <- debounce(
+          reactive({
+            req(input$dataname, input$x, input$y)
+            within(data(),
+              {
+                plot <- ggplot2::ggplot(dataname, ggplot2::aes(x = x, y = y)) +
+                  ggplot2::geom_point()
+              },
+              dataname = as.name(input$dataname),
+              x = as.name(input$x),
+              y = as.name(input$y)
+            )
+          }), 200
+        )
+
+        plot_data_decorated <- srv_transform_data("decorate", data = plot_data, transforms = decorators)
+
+        plot_r <- reactive({
+          req(plot_data_decorated())
+          plot_data_decorated()[["plot"]]
+        })
+
+        output$plot <- renderPlot(plot_r())
+        output$text <- renderText({
+          teal.code::get_code(req(plot_data_decorated()))
+        })
+      })
+    },
+    ui_args = list(decorators = decorators),
+    server_args = list(decorators = decorators)
+  )
+}
+```
+
+### Application
+
+```{r}
+library(ggplot2)
+app <- init(
+  data = teal_data(iris = iris, mtcars = mtcars),
+  modules = modules(
+    tm_decorated_plot("identity"),
+    tm_decorated_plot("no-ui", decorators = static_decorator),
+    tm_decorated_plot("lang", decorators = static_decorator_lang),
+    tm_decorated_plot("interactive", decorators = interactive_decorator),
+    tm_decorated_plot("interactive-from lang", decorators = interactive_decorator_lang),
+    tm_decorated_plot("from-fun", decorators = gg_xlab_decorator("plot")),
+    tm_decorated_plot("failing", decorators = failing_decorator)
+  )
+)
+
+if (interactive()) {
+  shinyApp(app$ui, app$server)
+}
+```
+
+## Decorating Tables
+
+### Example Module
+
+```{r, eval=requireNamespace("DT")}
+library(DT)
+
+# Define a decorator module for customizing table appearance
+custom_table_decorator <- teal_transform_module(
```

Comment:
I think `custom_table_decorator` should occupy separate chunk and be described

Reply:
f9bdb60c692ffc465975f747232eee164b36a2d9